### PR TITLE
Fixes #25406 - Update CV publish error to more friendly one.

### PR DESCRIPTION
**Pull request by chris1984**

### CODE REVIEW

**Comment by chris1984** [comment]
@beav here is the pr for the updated error message

**Comment by beav** [comment]
does this render correctly? I haven't tested it yet but it looks like it would render as `Minor, CV Version with same major and minor version already exists.`

**Comment by chris1984** [comment]
@beav good catch i will update it to fix it

**Comment by beav** [comment]
Also you may want to create a new issue instead of referring to the older one. At this point it's a new issue since the old one was already merged.

**Comment by chris1984** [comment]
[test katello]

**Comment by chris1984** [comment]
@beav this is ready for review again, I tested it on my dev env and this is what httpie returns when i try to pub the same cvv:

```
{
    "displayMessage": "Validation failed: Minor CV Version with same major and minor version already exists.",
    "errors": {
        "minor": [
            "CV Version with same major and minor version already exists."
        ]
    }
}
```

**Comment by chris1984** [comment]
Also updated the issue # to a new one

File: app/models/katello/content_view_version.rb [hunk 1]
```diff
@@ -46,7 +46,7 @@ class ContentViewVersion < Katello::Model
 
     validates_lengths_from_database
 
-    validates :minor, :uniqueness => {:scope => [:content_view_id, :major], :message => N_(", must be unique to major and version id version.")}
+    validates :minor, :uniqueness => {:scope => [:content_view_id, :major], :message => "CV Version with same major and minor version already exists."}
```
Comment (ehelms) [inline app/models/katello/content_view_version.rb]: I think we still want the internalization here (`N_(`)

Comment (beav) [reply app/models/katello/content_view_version.rb]: I haven't been able to look at this yet, but is there a way to keep rails from prepending 'Minor' to the message?

Comment (chris1984) [reply app/models/katello/content_view_version.rb]: @beav I tried a few things with rails 5 syntax and still no good, from things I tried manually or trying things from different articles. It shows the correct error message for the 2nd message but the first one is still shows `minor`

**Comment by beav** [comment]
@chris1984 this error may be better caught in the controller than the model. For example

```diff
diff --git a/app/controllers/katello/api/v2/content_views_controller.rb b/app/controllers/katello/api/v2/content_views_controller.rb
index 94dc70f..a1733ae 100644
--- a/app/controllers/katello/api/v2/content_views_controller.rb
+++ b/app/controllers/katello/api/v2/content_views_controller.rb
@@ -85,6 +85,10 @@ module Katello
       param :rpm_filenames, Array, of: String, :desc => N_("list of rpm filename strings to include in published version"), :required => true
     end
     def publish
+      if params[:major].present? && params[:minor].present? && ContentViewVersion.find_by(:content_view_id => params[:id], :major => params[:major], :minor => params[:minor]).present?
+        fail HttpErrors::BadRequest, _("A CV version already exists with the same major and minor version")
+      end
+
       if params[:repos_units].present? && @view.composite?
         fail HttpErrors::BadRequest, _("Directly setting package lists on composite content views is not allowed. Please " \
                                      "update the components, then re-publish the composite.")

```

File: app/controllers/katello/api/v2/content_views_controller.rb [hunk 2]
```diff
@@ -89,6 +89,11 @@ def publish
         fail HttpErrors::BadRequest, _("Directly setting package lists on composite content views is not allowed. Please " \
                                      "update the components, then re-publish the composite.")
       end
+      if params[:major].present? && params[:minor].present? && ContentViewVersion.find_by(:content_view_id => params[:id], :major => params[:major], :minor => params[:minor]).present?
+        major = params[:major]
+        minor = params[:minor]
+        fail HttpErrors::BadRequest, _("A CV version already exists with the same major and minor version (#{major}.#{minor})")
```
Comment (beav) [inline app/controllers/katello/api/v2/content_views_controller.rb]: I would do just `#{params[:major]}.#{params[:minor]}` so you dont have to do the 2 lines above

**Comment by chris1984** [comment]
@beav updated the pr to get rid of the variables and use Ruby format strings to make the minitest happy.
```
{
    "displayMessage": "A CV version already exists with the same major and minor version (4.1)",
    "errors": [
        "A CV version already exists with the same major and minor version (4.1)"
    ]
}
```

**Comment by beav** [comment]
It looks like a unit test that checks source files for correctness does not like that there's two `%s` in the same string. This is likely so people don't get the parameters swapped by mistake. Try this instead, which names the major + minor for interpoloation:

```
fail HttpErrors::BadRequest, _("A CV version already exists with the same major and minor version (%{major}.%{minor})") % {:major => params[:major], :minor => params[:minor]}
```

**Comment by beav** [comment]
you should probably also add a unit test. for example,

```diff
--- a/test/controllers/api/v2/content_views_controller_test.rb
+++ b/test/controllers/api/v2/content_views_controller_test.rb
@@ -101,6 +101,13 @@ module Katello
       post :publish, params: { :id => target_view.id, :major => 4, :minor => 1 }
     end
 
+    def test_publish_with_duplicated_version_params
+      target_view = ContentView.find(katello_content_views(:library_dev_view).id)
+      ContentViewVersion.expects(:find_by).with(:content_view_id=>target_view.id.to_s, :major=>88, :minor=>1).returns(["non-nil result"])
+      post :publish, params: { :id => target_view.id, :major => 88, :minor => 1 }
+      assert_response 400
+    end  
+
     def test_create_fail_without_organization_id
       post :create, params: { :name => "My View", :label => "My_View", :description => "Cool" }
 
```

Basically when your check calls `ContentViewVersion.find_by`, we want to return something that will trigger `.present?` to be true, so it falls into the if statement and raises the error

**Comment by chris1984** [comment]
@beav @ehelms how does this look?

**Comment by beav** [comment]
code looks good and tested well, ack thanks @chris1984 !

**Pull request merged**